kernel/src/hil/usb_hid.rs

```diff
@@ -0,0 +1,100 @@
+//! Interface for USB HID (Human Interface Device) class
+
+use crate::returncode::ReturnCode;
+
+/// Implement this trait and use `set_client()` in order to receive callbacks.
+pub trait Client<'a> {
+    /// Called when a packet is received.
+    /// This will return the buffer passed into `receive_buffer()` as well as
+    /// the endpoint where the data was received. The `len` argument contains
+    /// the length of the packet received. If the buffer length is smaller
+    /// then this length the buffer will only contain part of the packet and
+    /// `result` will contain indicate an `ESIZE` error.
+    fn packet_received(
+        &'a self,
+        result: ReturnCode,
+        buffer: &'static mut [u8; 64],
+        len: usize,
+        endpoint: usize,
+    );
+
+    /// Called when a packet has been finished transmitting.
+    /// This will return the buffer passed into `send_buffer()` as well as
+    /// the endpoint where the data was sent. The `len` argument contains
+    /// the number of bytes sent. If this is less then the size of the
+    /// buffer `result` will contain the `ESIZE` error.
+    fn packet_transmitted(
+        &'a self,
+        result: ReturnCode,
+        buffer: &'static mut [u8; 64],
+        len: usize,
+        endpoint: usize,
+    );
+
+    /// Called when checking if we can start a new receive operation.
+    /// Should return true if we are ready to receive and not currently
+    /// in the process of receiving anything. That is if we are currently
+    /// idle.
+    /// If there is an outstanding call to receive, a callback already
+    /// waiting to be called then this will return false.
+    fn can_receive(&'a self) -> bool;
+}
+
+pub trait UsbHid<'a> {
+    /// Sets the buffer to be sent and starts a send transaction.
+    /// On success returns the number of bytes that will be sent.
+    /// On failure returns an error code and the buffer passed in.
```

Comment:
How is the number of bytes that will be sent calculated?

Reply:
It should always be 64 for high speed or 8 for low speed devices

Reply:
64 is for full speed, right? I thought high speed was 1024? (not that I have ever heard of a high speed HID device)

Reply:
High speed is USB 2.0, the HID spec says:

> A transaction may be made up of multiple packets (token, data, handshake) but is limited in size to 8 bytes for low-speed devices and 64 bytes for high-speed devices.

Reply:
Oh, interesting, thanks